File: compose/cli/colors.py
import enum
import os

from ..const import IS_WINDOWS_PLATFORM

NAMES = [
    'grey',
    'red',
    'green',
    'yellow',
    'blue',
    'magenta',
    'cyan',
    'white'
]


@enum.unique
class AnsiMode(enum.Enum):
    """Enumeration for when to output ANSI colors."""
    NEVER = "never"
    ALWAYS = "always"
    AUTO = "auto"

    def use_ansi_codes(self, stream):
        if self is AnsiMode.ALWAYS:
            return True
        if self is AnsiMode.NEVER or os.environ.get('CLICOLOR') == '0':
            return False
        return stream.isatty()


def get_pairs():
    for i, name in enumerate(NAMES):
        yield (name, str(30 + i))
        yield ('intense_' + name, str(30 + i) + ';1')


def ansi(code):
    return '\033[{}m'.format(code)


def ansi_color(code, s):
    return '{}{}{}'.format(ansi(code), s, ansi(0))


def make_color_fn(code):
    return lambda s: ansi_color(code, s)


if IS_WINDOWS_PLATFORM:
    import colorama
    colorama.init(strip=False)
for (name, code) in get_pairs():
    globals()[name] = make_color_fn(code)


def rainbow():
    cs = ['cyan', 'yellow', 'green', 'magenta', 'blue',
          'intense_cyan', 'intense_yellow', 'intense_green',
          'intense_magenta', 'intense_blue']

    for c in cs:
        yield globals()[c]

File: ecs/pkg/console/colors.go
package console

import (
	"fmt"
	"strconv"
)

var NAMES = []string{
	"grey",
	"red",
	"green",
	"yellow",
	"blue",
	"magenta",
	"cyan",
	"white",
}

var COLORS map[string]ColorFunc

// ColorFunc use ANSI codes to render colored text on console
type ColorFunc func(s string) string

var Monochrome = func(s string) string {
	return s
}

func ansiColor(code, s string) string {
	return fmt.Sprintf("%s%s%s", ansi(code), s, ansi("0"))
}

func ansi(code string) string {
	return fmt.Sprintf("\033[%sm", code)
}

func makeColorFunc(code string) ColorFunc {
	return func(s string) string {
		return ansiColor(code, s)
	}
}

var Rainbow = make(chan ColorFunc)

func init() {
	COLORS = map[string]ColorFunc{}
	for i, name := range NAMES {
		COLORS[name] = makeColorFunc(strconv.Itoa(30 + i))
		COLORS["intense_"+name] = makeColorFunc(strconv.Itoa(30+i) + ";1")
	}

	go func() {
		i := 0
		rainbow := []ColorFunc{
			COLORS["cyan"],
			COLORS["yellow"],
			COLORS["green"],
			COLORS["magenta"],
			COLORS["blue"],
			COLORS["intense_cyan"],
			COLORS["intense_yellow"],
			COLORS["intense_green"],
			COLORS["intense_magenta"],
			COLORS["intense_blue"],
		}

		for {
			Rainbow <- rainbow[i]
			i = (i + 1) % len(rainbow)
		}
	}()
}
